Bug on Windows: game crashes when the PC wakes up from sleep

tyomitch opened the pull request:
If I close and reopen my laptop during the game, it crashes with

> SDL Error
> LockRect(): INVALIDCALL
> The error occurred at: ..\..\SourceX\dx.cpp line 264

I have absolutely no experience with SDL, but I've managed to put
together a workaround. Improvements are welcome.

AJenbo commented:
Interesting, ill try and find a laptop to test on :)

tyomitch commented:
> Interesting, ill try and find a laptop to test on :)

I don't currently have a desktop PC, but I suppose you can get the same effect there by pressing the power button during the game, to send the PC to sleep.

AJenbo commented:
I was unable to replicate this both on Linux, and windows 10.

Could you give some specs for you machine and sdl version your using?

tyomitch commented:
Fujitsu Lifebook E752, Windows 10 ver. 1909, SDL 2.0.10: https://youtu.be/obSPmm1D580 (sorry for the shaky phone!)

AJenbo commented:
Looks like it's an Intel GPU, the Windows machine I tested on had an AMD GPU. I do have a machine with an Intel GPU, but it doesn't have Windows installed.

Could you try SDL 2.0.12?

tyomitch commented:
> Looks like it's an Intel GPU

Yes, it's Intel HD Graphics 4000

> Could you try SDL 2.0.12?

Still the same :-(

AJenbo commented on SourceX/dx.cpp:
```diff
@@ -261,7 +261,31 @@ void RenderPresent()
 		if (SDL_UpdateTexture(texture, NULL, surface->pixels, surface->pitch) <= -1) { //pitch is 2560
-			ErrSdl();
+			// We should supposedly handle SDL_RENDER_DEVICE_RESET, but I tried and it doesn't arrive
+			if (!strcmp(SDL_GetError(), "LockRect(): INVALIDCALL")) {
```
I wonder if this could be detected similar to what was done with GetOutputSurface based on the state of the texture. could you check the state of SDL_WasInit() and see what returns?

tyomitch replied:
`0xe231` which is the same value as passed to `SDL_Init`

I guess that only the texture(s) get invalidated by the sleep, not the whole of SDL.

tyomitch commented:
Will it help if I upload a dump file, stopped at a breakpoint after the sleep, with the texture invalidated?

tyomitch committed:
Bug on Windows: game crashes when the PC wakes up from sleep

If I close and reopen my laptop during the game, it crashes with

> SDL Error
> LockRect(): INVALIDCALL
> The error occurred at: ..\..\Source\dx.cpp line 313

tyomitch committed:
Extract `ReinitializeRenderer()` into a function

Also, handle `SDL_RENDER_DEVICE_RESET`
